### ISSUE TITLE
Incorrect null warnings with null-check in recursive patterns

### ISSUE THREAD

**Issue by alrz** [dotnet/roslyn#44534]
**Version Used**:  287a25a

**Steps to Reproduce**:
```cs
#nullable enable
public class C {
    public void M1(object? a, object? b) {
        switch (a, b) 
        {
            case (null, _):
            case (_, null):
                return;
        }     
        a.ToString(); // warns (none expected)
        b.ToString(); // warns (none expected)       
    }
    
    public void M2(object? a, object? b) {
        if ((a, b) is (null, _) or (_, null)) 
            return;
        a.ToString(); // warns (none expected)
        b.ToString(); // warns (none expected)     
    }
    
    public void M3(object? a, object? b) {
        switch (a) { case null: return; }
        switch (b) { case null: return; }
        a.ToString(); // ok
        b.ToString(); // ok       
    }
}
```

([sharplab repro](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4GIB2BXANliwtwAEcaeBAsAFAACAzIdQEyEDChA3lYdw/dQCyEAsgAYAFAHtgAKzgBjGAH5CEADSEpshcuABKDlx5GoAdwCWMOQAtCYtXsNHunSk7eE50ImMw51AfV0QR3duTyhvf3VfLCCQ0KNqAHYAbnieAF8ndO4IADoAFQkAZRgEMzQAczFdNNd3YEKSsorq2vcQjJCQugZBIUZJGXklFXVNEZ19F1CzADNbO1U9QjMoWxiA/QkEWyjCGN19HKdkutD8otLyqprzhqbr1ruO+u4ut8Ievn7aIa1RmoNMNtIQVjN3KYLNZbBBph4vAdsFgQAxUoQPqEoZYbGJwQiIkicKizhiTpdmjc2vc3I0ri1bu03J0qBkgA=))

**Comment by alrz** [comment]
Another scenario:
```cs
    public void M4(object? a, object? b) {
        // either
        if ((a, b) is not ({}, {})) return;
        if (!((a, b) is not ({}, {}))) return;
        a.ToString(); // warns (none expected)
        b.ToString(); // warns (none expected)     
    }
```

**Comment by alrz** [comment]
Since we do recognize `(null or int)` or `not null` as a null check, it seems like this is only related to recursive patterns.

**Comment by jnm2** [comment]
Is this the same or should a new issue be opened?

Provided by @badcel:

```cs
using System;

namespace switchexpression
{
    public interface IData 
    {
        string? Text { get; }
    }
    public class Data : IData
    {
        public string? Text { get; set; }
    }
    class Program
    {
        static void Main(string[] args)
        {
            IData myData = new Data();
            string ret = myData switch
            {
                Data { Text: { } } d => d.Text, // possible null to non nullable type conversion error
                { Text: { } } => myData.Text, // OK
                _ => ""
            };
        }
    }
}
```

**Comment by alrz** [comment]
I imagine that's the same issue. `Data { Text: { } } d` is a recursive pattern that we don't cover. Note `{ Text: { } t }` already works for `t`, but any nested null state is lost.

**Comment by alrz** [comment]
that's strange though. See https://github.com/dotnet/roslyn/issues/42057#issuecomment-593547831 for an example where we do recognize it. So I'm not sure if that's the same issue.

**Comment by jcouv** [comment]
Verified the three scenarios described about were fixed by https://github.com/dotnet/roslyn/pull/51001 (merged for 16.10)
I'll close the issue. Thanks
